lib/Dialect/TritonGPU/Transforms/Pipeliner/SoftwarePipeliner.cpp

```diff
@@ -42,6 +43,102 @@ static void pipelineWgmma(ModuleOp moduleOp) {
   }
 }
 
+static Operation *wrapInMaskOp(RewriterBase &rewriter, Operation *op,
+                               Value pred) {
+  if (isa<arith::ConstantOp>(op)) {
+    return op;
+  }
+  auto mask = rewriter.create<MaskOp>(op->getLoc(), op->getResultTypes(), pred);
+  rewriter.createBlock(&mask->getRegion(0));
+  rewriter.setInsertionPointToStart(&mask->getRegion(0).front());
+  auto newOp = rewriter.clone(*op);
+  rewriter.create<MaskReturnOp>(op->getLoc(), newOp->getResults());
+  op->replaceAllUsesWith(mask->getResults());
+  rewriter.eraseOp(op);
+  return mask;
+}
+
+static void resolveMaskOp(ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp);
+
+  // Canonicalize the IR to simplify the arithmetic ops defining the mask
+  auto arithDialect =
+      moduleOp.getContext()->getLoadedDialect<arith::ArithDialect>();
+  RewritePatternSet patterns(moduleOp.getContext());
+  arithDialect->getCanonicalizationPatterns(patterns);
+  if (applyPatternsGreedily(moduleOp, std::move(patterns)).failed())
+    return llvm::report_fatal_error("Failed to canonicalize the IR");
+
+  SmallVector<MaskOp> maskOps;
+  moduleOp->walk([&](MaskOp maskOp) { maskOps.push_back(maskOp); });
+  for (auto maskOp : maskOps) {
+    rewriter.setInsertionPoint(maskOp);
+    while (&maskOp.getBody()->front() != maskOp.getBody()->getTerminator()) {
+      Operation *op = &maskOp.getBody()->front();
+      // Statically dead
+      if (isConstantIntValue(maskOp.getPred(), 0)) {
+        if (op->getNumResults() > 0) {
+          SmallVector<Value> results;
+          for (auto result : op->getResults()) {
+            auto poisonOp =
```

Comment:
Just a note: I have observed that using constant 0 sometimes results in significantly better codegen than `poison`. See https://github.com/triton-lang/triton/blob/75fe1135f4cbc7fefff02b6cf497d340fbc7d9cc/lib/Dialect/TritonGPU/Transforms/FuseNestedLoops.cpp#L292

Reply:
the poison should all be gone by the time we get to llvm